### ISSUE TITLE
Request a rights sample file

### ISSUE THREAD

**Issue by P-Kito** [Splamy/TS3AudioBot#433]
**Is your feature request related to a problem? Please describe.**
Easy as that, request a rights sample file, which is always accessable.
When coding a different software project, we used to have the config.cfg.dist file to always get updated aswell, so that you can see differences with all rights filled in + their default setting.
( https://github.com/TrinityCore/TrinityCore/blob/master/src/server/worldserver/worldserver.conf.dist )

**Describe the solution you'd like**
See above

**Describe alternatives you've considered**
Put a sample file onto the wiki and keep it updated.

**Why am I requesting?**
I recently found myself lacking some permissions (webinterface stuff) and I was wondering why they are not in the rights.toml.. so I tried to find out what permissions are new and why my file is fcked up....

I saw there is one in the src, but is it really up to date? I couldnt even find cmd.version on it..

**Comment by Splamy** [comment]
In what way do you consider accessible? Via console, or via command, or some other way?

The rights file in the audiobot source should alyways be up to date. It's the default file for a new setup after all.
The `cmd.version` for example isn't in there because it is not a command I considered being either useful fo the normal user, nor do I know if every selfhoter would want this command available for public.
So evey command which doesn't fall into either 'basic' or 'playing' permission are in the admin rule with the `= "*"`

Can you give me an example how you would use this? maybe I can get some more insight

**Comment by P-Kito** [comment]
> The rights file in the audiobot source should alyways be up to date. It's the default file for a new setup after all.

As soon as I did not find "cmd.version" I got suspicious... why is this config file not complete? Why is right definition such a clunky thing on TS3AudioBot... Maybe could be my own opinion, but I feel like permissions are weird on this project... or maybe just "toml" files.

I am probably just really unhappy about the solution with "toml" files. I still feel like a "casual" SQL (sqlite?) solution would be way better for generating M:N Relations (Permission:Servergroup)...

But having a "ready-to-use" rights.toml file, which you copy paste and put in your server groups, would really help.
A toml file, which shows everything, works immediatly and is ready-to-use for the webinterface without playing around or tinkering the file and probably breaking stuff.

Maybe you could make the default file a bit "better" and link it officially into the readme/setup.
_Why should it be added only when bot is setting up once?_ We need a **rights.toml.default file** in the Release dictionary at any time.... I feel like.

**Comment by Flakebi** [comment]
If you are a “power user” I agree that SQL may be more convenient. I think for a user who does not manage dozens of bots on multiple servers this would be overly complicated. It is much easier to open a text file and edit it than connecting to a database and look at the tables in there.

The ready-to-use rights.toml file which also gets used when you do not create the file before starting the bot (as far as I know) is [this one](https://github.com/Splamy/TS3AudioBot/blob/master/TS3AudioBot/Rights/DefaultRights.toml).
The commands which are explicitly defined at the top are allowed for every user.
Other commands like `!version` are only allowed for admins: `"+" = "*"`

Having this file as `rights.toml.default` or something like that sounds good to me.

One thing I am not sure about with SQL is how to implement complicated filters.
For example what tables would you need to allow server group 2 on server abc.net to access the !play command?
Or allow a specific user id to access all commands on every server?
Or anonymous api requests?

**Comment by P-Kito** [comment]
Something like... **BEWARE: 5-Minutes thoughts**

```sql
tableGroups:
   int groupId
   int permission (See Below)
   varchar comment/name

tableGroupsToTSGroup
   int serverGroupTS (-1 reserved for anonymous M:N)
   int serverId (or hostname?)
   int groupId (foreignkey)
   varchar comment
```

Having permissions in code like:

```cpp
enum Permissions 
{
   COMMAND_BOT_LIST   = 1,
   COMMAND_BOT_NAME   = 2,
   COMMAND_BOT_SOME   = 3,
   // Reserve space for more
   COMMAND_PLAY   = 50,
   COMMAND_STOP   = 51,
   // Reserve space for more
   COMMAND_HISTORY_ADD = 100,
   COMMAND_HISTORY_DEL = 101
}
```

```sql
-- Sort permissions into groups
INSERT INTO tableGroups VALUES (0, 1, "Bot Commands Group");
INSERT INTO tableGroups VALUES (0, 2, "Bot Commands Group");
INSERT INTO tableGroups VALUES (0, 3, "Bot Commands Group");
INSERT INTO tableGroups VALUES (1, 50, "Play Commands Group");
INSERT INTO tableGroups VALUES (1, 51, "Play Commands Group");
INSERT INTO tableGroups VALUES (2, 100, "History Commands Group");
INSERT INTO tableGroups VALUES (2, 101, "History Commands Group");

-- Sort permission-groups to ts-groups
INSERT INTO tableGroupsToTSGroup VALUES (6, 0, 0, "Servergroup 6 (Admin) Bot Commands");
INSERT INTO tableGroupsToTSGroup VALUES (6, 1, 0, "Servergroup 6 (Admin) Play Commands");
INSERT INTO tableGroupsToTSGroup VALUES (6, 2, 0, "Servergroup 6 (Admin) History Commands");
-- User Permissions
INSERT INTO tableGroupsToTSGroup VALUES (8, 1, 0, "Servergroup 8 (User) Play Commands");
INSERT INTO tableGroupsToTSGroup VALUES (8, 2, 0, "Servergroup 8 (User) History Commands");
-- API Calls (no server group)
INSERT INTO tableGroupsToTSGroup VALUES (-1, 1, 0, "Anonymous Calls (API) Play Commands");
```

**Comment by gabanzin** [comment]
Could you let me know where I put the permissions? Sorry, Im so _dumb_ and took 4 hour to make this bot run in my ubuntu vps. I just throwed it in "rights.toml" but when I started running the bot gave me error (sorry, I dont remember what kind of error).

**Comment by Splamy** [comment]
@gabanzin The bot should create a default rights.toml where you can simply set your uid or group id into the predefined fields.
There's also a wiki page with a few examples: https://github.com/Splamy/TS3AudioBot/wiki/Rights#examples
Furthermore, could you descript _what_ took you so long during the setup process? Maybe we can improve on that if we get some useful feedback from new users.

@P-Kito so, I took some time to answer as I had some busy days and I wanted to think about it.
I have realize two major (very correlated) problems for the most users with the permission configuration:
1. The "signal-to-noise-ratio" is very high. So basically speaking there are _a lot_ of permissions, but no convenient default groups.
2. For the user interface (api) the user doesn't know (and normally shouln't even have to) _which_ permissions are _required_ to properly use a given webinterface site (like home, bots, a bot instance, etc)

My solution to both of those would be to provide some useful default groups.
The rights-system already supports defining and using custon command groups.
I would provide some default groups like `api`, `web_home`, `web_bots`, ...
This would ease the backend setup, as you can include useful permission groups like this:
```toml
[[rule]]
bot = "default"
include = [ 'api', "web_home" ]
```
And the front-end (webinterface) would benefit from this by allowing simple operations like adding or removing such groups with a simple toggle.
Example from the sinusbot (ignore the red arrow; it's the first best image I found):
![image](https://user-images.githubusercontent.com/6502527/49212009-45cb7700-f3c1-11e8-83cd-adf62a189237.png)

As for your SQL solution; I think it's quite a bit ironic that this doesn't even sole your own problem.
> I tried to find out what permissions are new

Your problem fell into my category 2. If the permission system was handled in a SQL database you would be in the exact same position that you'd have to scrap through either the source code or database to search for what you actully need/want.

Furthermore there are a lot more problems comming up with this. For example currently thosse permissions are automatically generated by the command name, which means, there is a easy conversion from commands name to permission; wich is very convenient for plugins where commands can be added or removed at runtime any time.

Your SQL example has a lot less configurability. Matcher like, per bot, per uid, per api, ... are missing. For example I like to match by uid, because it's generally a lot faster and more reliable across servers where bots don't have that much ts3permissions. Logical combinations from rule are also mission..
Replicating such funtionality in a database does not seem to be trivial.

**Comment by gabanzin** [comment]
@Splamy Sure, my biggest obstacle was how to exec the bot, you can put in "readme.md" that to open the bot you need to go "TS3AudioBot/TS3AudioBot/bin/Release/net64"... I also have difficult to build the bot "msbuild", had some kind of error... But a little of research in issue I fixed it.

---OFF TOPIC---
I gave up of the bot because my teamsepak was lagging... My vpn configs:
ram: 512mb
network: 50 Mbit/s
cpu: 1 vCPU of Intel Xeon E5-2630 V4 (I dont know what is this, but the processor never exceeds 10%.

I think is RAM... Sorry about bad english

**Comment by Splamy** [comment]
Did you know, that you don't have to build the bot manually? There are nightly build which are ready to download and run (with mono)

**Comment by gabanzin** [comment]
@Splamy yep, I followed the tutorial with mono, and I remember I get failed... So I proceeded to build manually... I just running teamspeak in my vps.

**Comment by Splamy** [comment]
Closing in favor of the tracking issue #651
